### PR TITLE
Withdraw testset code refactored

### PR DESCRIPTION
**Pull request by artemwoofsoftware**

### CODE REVIEW

File: test/withdraw-test.ts [hunk 1]
```diff
@@ -1,486 +1,644 @@
...
-    expect(event(s0, 1)).to.be.deep.equal({
-      Withdraw: {
-        src: bob.address,
-        to: alice.address,
-        amount: 0n,
-      }
+  afterEach(async () => {
+    await baseSnapshot.restore();
+    baseSnapshot = await takeSnapshot();
+  });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: We can do without creating a snapshot

File: test/withdraw-test.ts [hunk 2]
```diff
@@ -1,486 +1,644 @@
...
+
+        await expect(comet.connect(alice).withdraw(baseToken.address, 1)).to.be.revertedWithCustomError(comet, 'Paused');
+        await comet.connect(pauseGuardian).pause(false, false, false, false, false);
+      });
+
+      it('reverts if withdrawing base exceeds the total supply (harness - artificial state)', async () => {
+        // This tests an edge case where balance > actual tokens in contract
+        // Must use harness to create this inconsistent state
+        await baseToken.allocateTo(comet.address, 100e6);
+        await comet.setBasePrincipal(bob.address, 100e6);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: Remove `setBasePrincipal` function and make a real flow with supplying tokens

File: test/withdraw-test.ts [hunk 3]
```diff
@@ -1,486 +1,644 @@
...
+        expect(await comet.isWithdrawPaused()).to.be.true;
+
+        await expect(comet.connect(alice).withdraw(baseToken.address, 1)).to.be.revertedWithCustomError(comet, 'Paused');
+        await comet.connect(pauseGuardian).pause(false, false, false, false, false);
+      });
+
+      it('reverts if withdrawing base exceeds the total supply (harness - artificial state)', async () => {
+        // This tests an edge case where balance > actual tokens in contract
+        // Must use harness to create this inconsistent state
+        await baseToken.allocateTo(comet.address, 100e6);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: We can allocate tokens in `before` block before describe block revert, and 1 allocation will work for happy path and revert cases

File: test/withdraw-test.ts [hunk 4]
```diff
@@ -1,486 +1,644 @@
...
+        await comet.setBasePrincipal(bob.address, 100e6);
+
+        await expect(comet.connect(bob).withdraw(baseToken.address, 100e6)).to.be.reverted;
+      });
+
+      it('reverts if the asset is neither collateral nor base', async () => {
+        await unsupportedToken.allocateTo(comet.address, 1);
+
+        await expect(comet.connect(bob).withdraw(unsupportedToken.address, 1)).to.be.reverted;
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: This test can be moved to the global describe block `edge cases` as it is not related to the base token and collaterals

File: test/withdraw-test.ts [hunk 5]
```diff
@@ -1,486 +1,644 @@
...
+
+      it('reverts if withdraw max for a collateral asset', async () => {
+        const collateral = collaterals['COMP'];
+        await collateral.allocateTo(bob.address, 100e6);
+
+        await expect(
+          comet.connect(bob).withdraw(collateral.address, ethers.constants.MaxUint256)
+        ).to.be.revertedWith("custom error 'InvalidUInt128()'");
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:100]: Check for additional scenarios such as: BorrowToSmall, NotCollateralized

File: test/withdraw-test.ts [hunk 6]
```diff
@@ -1,486 +1,644 @@
...
+
+        const bobBalanceBefore = await baseToken.balanceOf(bob.address);
+        const aliceBalanceBefore = await baseToken.balanceOf(alice.address);
+
+        const withdrawTx = await comet.connect(bob).withdrawTo(alice.address, baseToken.address, SUPPLY_AMOUNT);
+        const receipt = await withdrawTx.wait();
+
+        expect(await comet.balanceOf(bob.address)).to.equal(0);
+        expect(await baseToken.balanceOf(alice.address)).to.equal(aliceBalanceBefore.add(SUPPLY_AMOUNT));
+        expect(await baseToken.balanceOf(bob.address)).to.equal(bobBalanceBefore);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: each of the check move to separate tests before withdraw tx, then preform withdraw and then in each separate test for correct changes

File: test/withdraw-test.ts [hunk 7]
```diff
@@ -1,486 +1,644 @@
...
+    describe('withdraw base: happy path', function () {
+      const SUPPLY_AMOUNT: bigint = exp(100, baseTokenDecimals);
 
-  // This demonstrates a weird quirk of the present value/principal value rounding down math.
-  it('withdraws 0 but Comet Transfer event amount is 1', async () => {
-    const protocol = await makeProtocol({ base: 'USDC' });
-    const { comet, tokens, users: [alice] } = protocol;
-    const { USDC } = tokens;
+      it('emits Transfer event (ERC20) when withdrawing base', async () => {
+        const snapshot = await takeSnapshot();
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: remove snapshot

File: test/withdraw-test.ts [hunk 8]
```diff
@@ -1,486 +1,644 @@
...
+
+        await baseToken.connect(bob).approve(comet.address, SUPPLY_AMOUNT);
+        await comet.connect(bob).supply(baseToken.address, SUPPLY_AMOUNT);
+
+        await expect(comet.connect(bob).withdrawTo(alice.address, baseToken.address, SUPPLY_AMOUNT))
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: move this tx in separate variable e.g. withdrawTx
and check that it is not failing

File: test/withdraw-test.ts [hunk 9]
```diff
@@ -1,486 +1,644 @@
...
+          .to.emit(baseToken, 'Transfer')
+          .withArgs(comet.address, alice.address, SUPPLY_AMOUNT);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: in separate test check that withdrawTx is emitting this event
same fo for the rest of events

by doing this we can remove snapshots here at all

File: test/withdraw-test.ts [hunk 10]
```diff
@@ -1,486 +1,644 @@
...
+        const aliceBalanceBefore = await baseToken.balanceOf(alice.address);
+
+        const withdrawTx = await comet.connect(bob).withdrawTo(alice.address, baseToken.address, SUPPLY_AMOUNT);
+        const receipt = await withdrawTx.wait();
+
+        expect(await comet.balanceOf(bob.address)).to.equal(0);
+        expect(await baseToken.balanceOf(alice.address)).to.equal(aliceBalanceBefore.add(SUPPLY_AMOUNT));
+        expect(await baseToken.balanceOf(bob.address)).to.equal(bobBalanceBefore);
+
+        const t1 = await comet.totalsBasic();
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: t1 is not informational naming for a variable

File: test/withdraw-test.ts [hunk 11]
```diff
@@ -1,486 +1,644 @@
...
+        expect(t1.totalSupplyBase).to.equal(0n);
+        expect(t1.totalBorrowBase).to.equal(0n);
+        expect(Number(receipt.gasUsed)).to.be.lessThan(106000);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:188]: gasUsed tests should be seoarate test

File: test/withdraw-test.ts [hunk 12]
```diff
@@ -1,486 +1,644 @@
...
+      });
     });
 
-    const s0 = await wait(comet.connect(alice).withdraw(USDC.address, 0));
+    describe('withdraw max base', function () {
+      it('withdraws max base balance (including accrued) from sender', async () => {
+        await baseToken.connect(bob).approve(comet.address, 100e6);
+        await comet.connect(bob).supply(baseToken.address, 100e6);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: by withdrawing some part of supply in previous tests, we can avoid supplying again for such flow

File: test/withdraw-test.ts [hunk 13]
```diff
@@ -1,486 +1,644 @@
...
+
+        await collaterals['WETH'].allocateTo(alice.address, exp(10, 18));
+        await collaterals['WETH'].connect(alice).approve(comet.address, exp(10, 18));
+        await comet.connect(alice).supply(collaterals['WETH'].address, exp(10, 18));
+        await comet.connect(alice).withdraw(baseToken.address, 50e6);
+
+        await baseToken.allocateTo(comet.address, 60e6);
+
+        await fastForward(86400);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: in comments we can add some info, how much exactly we skip

File: test/withdraw-test.ts [hunk 14]
```diff
@@ -1,486 +1,644 @@
...
-      Transfer: {
-        from: comet.address,
-        to: alice.address,
-        amount: 0n,
-      }
+        const aliceUsdcBefore = await baseToken.balanceOf(alice.address);
+        await comet.connect(bob).withdrawTo(alice.address, baseToken.address, ethers.constants.MaxUint256);
+
+        expect(await comet.balanceOf(bob.address)).to.equal(0);
+        expect(await baseToken.balanceOf(alice.address)).to.equal(aliceUsdcBefore.add(bobAccruedBalance));
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: separate tests

In final view we should have strcuture like:
- tests for state before 
- actual withdraw tx
- emitted or not emitted events with args, and why this args has such values
- state after

File: test/withdraw-test.ts [hunk 15]
```diff
@@ -1,486 +1,644 @@
...
+        expect(event({ receipt }, 0)).to.be.deep.equal({
+          Transfer: { from: comet.address, to: alice.address, amount: 0n }
+        });
+        expect(event({ receipt }, 1)).to.be.deep.equal({
+          Withdraw: { src: bob.address, to: alice.address, amount: 0n }
+        });
+
+        expect(await baseToken.balanceOf(alice.address)).to.equal(charlieUsdcBefore);
+        expect(Number(receipt.gasUsed)).to.be.lessThan(121000);
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: move to block withdraw -> lenders -> edge cases
make a separate describe block and make a structure described in the previous comment

File: test/withdraw-test.ts [hunk 16]
```diff
@@ -1,486 +1,644 @@
...
+        const balanceAfterAccrual = (await comet.callStatic.balanceOf(bob.address)).toBigInt();
+        expect(balanceAfterAccrual).to.be.gte(supplyAmount);
+
+        await baseToken.allocateTo(alice.address, exp(60, 6));
+        await baseToken.connect(alice).approve(comet.address, exp(60, 6));
+        await comet.connect(alice).supply(baseToken.address, exp(60, 6));
+
+        await comet.connect(bob).withdraw(baseToken.address, balanceAfterAccrual);
+        const finalBalance = await comet.callStatic.balanceOf(bob.address);
+        expect(finalBalance).to.be.equal(0);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: same here

File: test/withdraw-test.ts [hunk 17]
```diff
@@ -1,486 +1,644 @@
...
-        src: alice.address,
-        to: alice.address,
-        amount: 0n,
-      }
+
+    describe('withdraw with index calculation', function () {
+      it('calculates base principal correctly (index math test)', async () => {
+        // Harness: need exact index=2e15 to test principal*index math, can't achieve naturally
+        await baseToken.allocateTo(comet.address, 100e6);
+        await setTotalsBasic(comet, {
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: remove setTotalsBasic

File: test/withdraw-test.ts [hunk 18]
```diff
@@ -1,486 +1,644 @@
...
+          baseSupplyIndex: 2e15,
+          totalSupplyBase: 50e6,
+        });
+        await comet.setBasePrincipal(bob.address, 50e6);
+
+        expect(await comet.balanceOf(bob.address)).to.equal(exp(100, 6));
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: exp(100, 6) can be move to constant

File: test/withdraw-test.ts [hunk 19]
```diff
@@ -1,486 +1,644 @@
...
+        const tx = await comet.connect(bob).withdrawTo(alice.address, baseToken.address, exp(1, 6));
+        const receipt = await tx.wait();
+
+        expect(receipt.events.length).to.be.equal(2);
+        expect(event({ receipt }, 0)).to.be.deep.equal({
+          Transfer: { from: comet.address, to: alice.address, amount: exp(1, 6) }
+        });
+        expect(event({ receipt }, 1)).to.be.deep.equal({
+          Withdraw: { src: bob.address, to: alice.address, amount: exp(1, 6) }
+        });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: we can simplify this by:
await expect(comet.connect(bob).withdrawTo(alice.address, baseToken.address, exp(1, 6)))
.to.emit(comet, 'Transfer').withArgs...
.to.emit(comet, 'Withdraw').withArgs...

**Review by VladyslavPerederWoof — CHANGES_REQUESTED** [review]

File: test/withdraw-test.ts [hunk 20]
```diff
@@ -1,635 +1,1060 @@
...
+      it('reverts if withdraw is paused', async () => {
+        await comet.connect(pauseGuardian).pause(false, false, true, false, false);
+        expect(await comet.isWithdrawPaused()).to.be.true;
+
+        await expect(comet.connect(alice).withdraw(baseToken.address, 1)).to.be.revertedWithCustomError(comet, 'Paused');
+        await comet.connect(pauseGuardian).pause(false, false, false, false, false);
+      });
+
+      it('reverts if withdrawing base exceeds the total supply (harness - artificial state)', async () => {
+        await baseSnapshot.restore();
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: do nothing, can be removed

File: test/withdraw-test.ts [hunk 21]
```diff
@@ -1,635 +1,1060 @@
...
+        
+        // This tests an edge case where balance > actual tokens in contract
+        // Must use harness to create this inconsistent state
+        await baseToken.allocateTo(comet.address, ARTIFICIAL_BALANCE);
+        await comet.setBasePrincipal(bob.address, ARTIFICIAL_BALANCE);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: instead of setting principal and make allocation, we can move such test to separate describe block, a it is a good edge case, that can show a situation when:
1st user supply tokens
2nd user make collateral supply and make a borrow
1st user can withdraw his suppied amount

Comment (VladyslavPerederWoof) [reply test/withdraw-test.ts]: we need to reproduce a real situations

File: test/withdraw-test.ts [hunk 22]
```diff
@@ -1,635 +1,1060 @@
...
+
+        await expect(comet.connect(bob).withdraw(baseToken.address, ARTIFICIAL_BALANCE)).to.be.reverted;
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: expecting a revert is not enough, need to check exaclty error message to make sure we got error what we expect

File: test/withdraw-test.ts [hunk 23]
```diff
@@ -1,635 +1,1060 @@
...
+      before(async () => {
+        await baseSnapshot.restore();
+        
+        await baseToken.connect(bob).approve(comet.address, SUPPLY_AMOUNT);
+        await comet.connect(bob).supply(baseToken.address, SUPPLY_AMOUNT);
+
+        bobBalanceBefore = (await baseToken.balanceOf(bob.address)).toBigInt();
+        aliceBalanceBefore = (await baseToken.balanceOf(alice.address)).toBigInt();
+
+        withdrawTx = await comet.connect(bob).withdrawTo(alice.address, baseToken.address, SUPPLY_AMOUNT);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: why withdrawTo? withdrawTo/From should be separate test set

File: test/withdraw-test.ts [hunk 24]
```diff
@@ -1,635 +1,1060 @@
...
+      it('reverts if withdraw max for a collateral asset', async () => {
+        await baseSnapshot.restore();
+        
+        const collateral = collaterals['COMP'];
+        await collateral.allocateTo(bob.address, COLLATERAL_AMOUNT);
+
+        await expect(
+          comet.connect(bob).withdraw(collateral.address, ethers.constants.MaxUint256)
+        ).to.be.revertedWithCustomError(comet, 'InvalidUInt128');
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:88]: better way:
        
```ts
        const snapshot = await takeSnapshot();
        
        const collateral = collaterals['COMP'];
        await collateral.allocateTo(bob.address, COLLATERAL_AMOUNT);

        await expect(
          comet.connect(bob).withdraw(collateral.address, ethers.constants.MaxUint256)
        ).to.be.revertedWithCustomError(comet, 'InvalidUInt128');

        await snapshot.restore();
```

File: test/withdraw-test.ts [hunk 25]
```diff
@@ -1,635 +1,1060 @@
...
+
+      it('reverts if asset is neither collateral nor base', async () => {
+        await baseSnapshot.restore();
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: redundant

File: test/withdraw-test.ts [hunk 26]
```diff
@@ -1,635 +1,1060 @@
...
+        
+        // Note: For unsupported tokens, the revert may be due to arithmetic underflow
+        // (when subtracting withdrawal amount from zero balance) rather than BadAsset error
+        await expect(
+          comet.connect(alice).withdraw(unsupportedToken.address, 1)
+        ).to.be.reverted;
+      });
+
+      it('reverts if withdraw amount exceeds uint104 max', async () => {
+        await baseSnapshot.restore();
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: redundant

File: test/withdraw-test.ts [hunk 27]
```diff
@@ -1,635 +1,1060 @@
...
-      }
+
+    describe('withdraw base: happy path', function () {
+      const SUPPLY_AMOUNT: bigint = exp(100, baseTokenDecimals);
+      let withdrawTx: ContractTransaction;
+      let bobBalanceBefore: bigint;
+      let aliceBalanceBefore: bigint;
+
+      before(async () => {
+        await baseSnapshot.restore();
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:128]: redundant, revert describe block must not affect the state

File: test/withdraw-test.ts [hunk 28]
```diff
@@ -1,635 +1,1060 @@
...
+        
+        await baseToken.connect(bob).approve(comet.address, SUPPLY_AMOUNT);
+        await comet.connect(bob).supply(baseToken.address, SUPPLY_AMOUNT);
+
+        bobBalanceBefore = (await baseToken.balanceOf(bob.address)).toBigInt();
+        aliceBalanceBefore = (await baseToken.balanceOf(alice.address)).toBigInt();
+
+        withdrawTx = await comet.connect(bob).withdrawTo(alice.address, baseToken.address, SUPPLY_AMOUNT);
+      });
+
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:147]: No checks for state before, just saving data.
Also move saving withdrawTx to a variable
Example: 

Image

Comment (VladyslavPerederWoof) [reply test/withdraw-test.ts:147]: states and balances that you check after also should be checked before

File: test/withdraw-test.ts [hunk 29]
```diff
@@ -1,635 +1,1060 @@
...
+        ).to.be.reverted;
+      });
+
+      it('reverts if withdraw amount exceeds uint104 max', async () => {
+        await baseSnapshot.restore();
+
+        // uint104 max = 2^104 - 1
+        const UINT104_MAX = 2n ** 104n - 1n;
+
+        await expect(
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:97]: should expect the exact error

File: test/withdraw-test.ts [hunk 30]
```diff
@@ -1,635 +1,1060 @@
...
+      });
+
+      it('reverts if asset is neither collateral nor base', async () => {
+        await baseSnapshot.restore();
+        
+        // Note: For unsupported tokens, the revert may be due to arithmetic underflow
+        // (when subtracting withdrawal amount from zero balance) rather than BadAsset error
+        await expect(
+          comet.connect(alice).withdraw(unsupportedToken.address, 1)
+        ).to.be.reverted;
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: should expect the exact error

File: test/withdraw-test.ts [hunk 31]
```diff
@@ -1,635 +1,1060 @@
...
+        await expect(withdrawTx).to.not.be.reverted;
+      });
+
+      it('bob comet balance is zero after max withdrawal', async () => {
+        expect(await comet.balanceOf(bob.address)).to.equal(0);
+      });
+
+      it('alice receives full accrued balance', async () => {
+        expect(await baseToken.balanceOf(alice.address)).to.equal(aliceBalanceBefore + bobAccruedBalance);
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:305]: good describe block. but not enough checks. As in comments before 
tests for state before, all stuff that gonna be changed or not (but related to) after tx -> then tx -> state after that have been chaged or not

**Review by VladyslavPerederWoof — CHANGES_REQUESTED** [review]

File: test/withdraw-test.ts [hunk 32]
```diff
@@ -1,635 +1,1060 @@
...
+        const tx = await comet.connect(bob).withdrawTo(alice.address, baseToken.address, BORROW_AMOUNT);
+        const receipt = await tx.wait();
+
+        expect(receipt.events.length).to.be.equal(2);
+        expect(event({ receipt }, 0)).to.be.deep.equal({
+          Transfer: { from: comet.address, to: alice.address, amount: BORROW_AMOUNT }
+        });
+        expect(event({ receipt }, 1)).to.be.deep.equal({
+          Withdraw: { src: bob.address, to: alice.address, amount: BORROW_AMOUNT }
+        });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: 1 test for tx
1 for 1 event
1 tests for other event

split it pls

**Review by VladyslavPerederWoof — CHANGES_REQUESTED** [review]

File: test/withdraw-test.ts [hunk 33]
```diff
@@ -1,635 +1,1078 @@
...
+
+        await collaterals['WETH'].allocateTo(alice.address, ALICE_COLLATERAL_AMOUNT);
+        await collaterals['WETH'].connect(alice).approve(comet.address, ALICE_COLLATERAL_AMOUNT);
+        await comet.connect(alice).supply(collaterals['WETH'].address, ALICE_COLLATERAL_AMOUNT);
+        await comet.connect(alice).withdraw(baseToken.address, ALICE_BORROW_AMOUNT);
+
+        await baseToken.allocateTo(comet.address, exp(60, 6));
+
+        await fastForward(TIME_FORWARD_SECONDS);
+        await ethers.provider.send('evm_mine', []);
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: Code dupblication at least 3 times. 
Think about creating describe block for max withdraw + full accrued balance. In describe block can be created a snapshot and before block with selected dublication. 
If need additional manupulations in before, for each describe can be added lines such as: line 198-201 or 240 - 242

File: test/withdraw-test.ts [hunk 34]
```diff
@@ -1,635 +1,1078 @@
...
+        await WETH.connect(alice).approve(comet.address, exp(2, 18));
+        await comet.connect(alice).supply(WETH.address, exp(2, 18));
+        await comet.connect(alice).withdraw(USDC.address, exp(100, 6));
+
+        await priceFeeds.WETH.setRoundData(1, 0, 0, 0, 1);
+
+        await expect(
+          comet.connect(alice).withdraw(WETH.address, exp(1, 18))
+        ).to.be.revertedWithCustomError(comet, 'BadPrice');
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: Why create a new protocol?
Use the one that already exists, just return the original price to the oracle at the end of the test.

File: test/withdraw-test.ts [hunk 35]
```diff
@@ -1,635 +1,1078 @@
...
+        await WETH.connect(alice).approve(comet.address, exp(2, 18));
+        await comet.connect(alice).supply(WETH.address, exp(2, 18));
+        await comet.connect(alice).withdraw(USDC.address, exp(100, 6)); // borrow
+
+        await priceFeeds.USDC.setRoundData(1, 0, 0, 0, 1);
+
+        await expect(
+          comet.connect(alice).withdraw(WETH.address, exp(1, 18))
+        ).to.be.revertedWithCustomError(comet, 'BadPrice');
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: Why create a new protocol?
Use the one that already exists, just return the original price to the oracle at the end of the test.

File: test/withdraw-test.ts [hunk 36]
```diff
@@ -1,635 +1,1078 @@
...
 
-describe('withdraw', function () {
-  it('withdraws to sender by default', async () => {
-    const protocol = await makeProtocol({ base: 'USDC' });
-    const { comet, tokens, users: [bob] } = protocol;
-    const { USDC } = tokens;
+      it("can't borrow less than minBorrow", async () => {
+        const protocol = await makeProtocol({ base: 'USDC', baseBorrowMin: exp(1, 6) });
+        const { comet, tokens, users: [alice, bob] } = protocol;
+        const USDC = tokens.USDC as FaucetToken;
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: Why create new protocol?

Take the value of baseBorrowMin-1 and make withdraw

File: test/withdraw-test.ts [hunk 37]
```diff
@@ -1,635 +1,1078 @@
...
+        await WETH.allocateTo(alice.address, exp(1, 18));
+        await WETH.connect(alice).approve(comet.address, exp(1, 18));
+        await comet.connect(alice).supply(WETH.address, exp(1, 18));
+
+        await priceFeeds.WETH.setRoundData(1, 0, 0, 0, 1);
+
+        await expect(
+          comet.connect(alice).withdraw(USDC.address, exp(1, 6))
+        ).to.be.revertedWithCustomError(comet, 'BadPrice');
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: The problem is described before

File: test/withdraw-test.ts [hunk 38]
```diff
@@ -1,635 +1,1078 @@
...
+        await WETH.allocateTo(alice.address, exp(1, 18));
+        await WETH.connect(alice).approve(comet.address, exp(1, 18));
+        await comet.connect(alice).supply(WETH.address, exp(1, 18));
+
+        await priceFeeds.USDC.setRoundData(1, 0, 0, 0, 1);
+
+        await expect(
+          comet.connect(alice).withdraw(USDC.address, exp(1, 6))
+        ).to.be.revertedWithCustomError(comet, 'BadPrice');
+      });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts]: The problem is described before

File: test/withdraw-test.ts [hunk 39]
```diff
@@ -1,635 +1,1078 @@
...
+
+        const borrowAmount = exp(100, 6);
+        const aliceBalanceBefore = await baseToken.balanceOf(alice.address);
+
+        await comet.connect(alice).withdraw(baseToken.address, borrowAmount);
+
+        expect(await baseToken.balanceOf(alice.address)).to.equal(aliceBalanceBefore.add(borrowAmount));
+        expect(await baseBalanceOf(comet as unknown as CometHarnessInterface, alice.address)).to.equal(BigInt(-borrowAmount));
+      });
+    });
```
Comment (VladyslavPerederWoof) [inline test/withdraw-test.ts:1476]: 24 collateral test need to check that all 24 collaterals can be withdraw from the comet back

Thus, you need to supply 24 collaterals first, then make checks, then make a withdraw each collateral and check that balances are right (comet, userCollateral, token's balanceOf etc)

Example: https://github.com/woof-software/comet/blob/woof-software/supply-tests/test/supply-test.ts#L1186-L1400

**Review by VladyslavPerederWoof — CHANGES_REQUESTED** [review]
